File: expansion_board/led/led.py
###############################################
#
#  file: led.py
#  update: 2024-08-10
#  usage: 
#      sudo python led.py
#
###############################################

import time
import gpiod

class GPIO:
    global gpio, gpiochip

    def __init__(self, gpionum, gpiochipx, val):
        self.gpiochip = gpiod.Chip(gpiochipx, gpiod.Chip.OPEN_BY_NUMBER)
        self.gpio = self.gpiochip.get_line(gpionum)
        self.gpio.request(consumer="gpio", type=gpiod.LINE_REQ_DIR_OUT, default_vals=[val])
    
    def set(self, val):
        self.gpio.set_value(val)

    def release(self):
        self.gpio.release()

# led gpiochip
gpiochip_led = "6"

# led gpionum
gpionum_r_led = 0
gpionum_g_led = 1
gpionum_b_led = 2

# led init
red_led = GPIO(gpionum_r_led, gpiochip_led, 1)          # 初始化红色led, 初始电平为高电平，灯灭
green_led = GPIO(gpionum_g_led, gpiochip_led, 1)        # 初始化绿色led, 初始电平为高电平，灯灭
blue_led = GPIO(gpionum_b_led, gpiochip_led, 1)         # 初始化蓝色led, 初始电平为高电平，灯灭

def main():
    try:
        while True:
            
            red_led.set(0)
            green_led.set(0)
            blue_led.set(0)

            time.sleep(0.5)

            red_led.set(1)
            green_led.set(1)
            blue_led.set(1)

            time.sleep(0.5)

    except Exception as e:
        
        print("exit...：", e)
    
    finally:
        
        # led off
        red_led.set(1)
        green_led.set(1)
        blue_led.set(1)

        # led release
        red_led.release()
        green_led.release()
        blue_led.release()
        
if __name__ == "__main__":  
    main()

File: expansion_board/led/led.c
/*
*
*   file: led.c
*   update: 2024-08-07
*   usage: 
*       sudo gcc -o led led.c -lgpiod
*       sudo ./led
*
*/

#include <gpiod.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h> 

#define GPIONUM_R_LED  (0)
#define GPIONUM_G_LED  (1)
#define GPIONUM_B_LED  (2)

struct gpiod_chip *led_gpiochip;        
struct gpiod_line *r_led_line;          
struct gpiod_line *g_led_line;       
struct gpiod_line *b_led_line;

static void sigint_handler(int sig_num) 
{    
    /* led off */
    gpiod_line_set_value(r_led_line, 1);
    gpiod_line_set_value(g_led_line, 1);
    gpiod_line_set_value(b_led_line, 1);

    /* release line */
    gpiod_line_release(r_led_line);
    gpiod_line_release(g_led_line);
    gpiod_line_release(b_led_line);

    gpiod_chip_close(led_gpiochip);

    exit(0);  
}

int main(int argc, char **argv)
{
    int ret;

    /* register exit signal ( Ctrl + c ) */
    signal(SIGINT, sigint_handler);

    /* get gpio controller */
    led_gpiochip = gpiod_chip_open("/dev/gpiochip6");  
    if(led_gpiochip == NULL)
    {
        printf("gpiod_chip_open error\n");
        return -1;
    }

    /* get gpio line */
    /* red led */
    r_led_line = gpiod_chip_get_line(led_gpiochip, GPIONUM_R_LED);
    if(r_led_line == NULL)
    {
        printf("gpiod_chip_get_line error : 0\n");
        return -1;
    }

    /* green led */
    g_led_line = gpiod_chip_get_line(led_gpiochip, GPIONUM_G_LED);
    if(g_led_line == NULL)
    {
        printf("gpiod_chip_get_line error : 1\n");
        return -1;
    }

    /* blue led */
    b_led_line = gpiod_chip_get_line(led_gpiochip, GPIONUM_B_LED);
    if(b_led_line == NULL)
    {
        printf("gpiod_chip_get_line error : 2\n");
        return -1;
    }

    /* set the line direction to output mode, and the initial level is high */
    /* red led */
    ret = gpiod_line_request_output(r_led_line, "r_led_line", 1);   
    if(ret < 0)
    {
        printf("gpiod_line_request_output error : r_led_line\n");
        return -1;
    }

    /* green led */
    ret = gpiod_line_request_output(g_led_line, "g_led_line", 1);   
    if(ret < 0)
    {
        printf("gpiod_line_request_output error : g_led_line\n");
        return -1;
    }

    /* blue led */
    ret = gpiod_line_request_output(b_led_line, "b_led_line", 1);   
    if(ret < 0)
    {
        printf("gpiod_line_request_output error : b_led_line\n");
        return -1;
    }

    while(1)
    {   
        /* led off */
        gpiod_line_set_value(r_led_line, 1);
        gpiod_line_set_value(g_led_line, 1);
        gpiod_line_set_value(b_led_line, 1);

        /* delay 0.5s */
        usleep(500000);

        /* led on */
        gpiod_line_set_value(r_led_line, 0);
        gpiod_line_set_value(g_led_line, 0);
        gpiod_line_set_value(b_led_line, 0);

        usleep(500000);
    }

    return 0;
}